Add support for Raspberry Pi 3 (Cortex-A53) at build time (v18)

graysky2 opened the pull request:
## Description
Support Raspberry Pi 3 (RPi3) with Cortex-A53 processor.

## Motivation and Context
Allow for the generation of hardware-specific optimized code through gcc supported hardware-specific flags.  This is no different in principal to what is currently shipping (ie RPi and RPi2 specific options).

## How Has This Been Tested?
Minimal testing on my part so far.  Builds and runs fine.  No regressions using Kodi to play different formatted video and audio files.  I did not conduct any controlled benchmarks, but since the RPi is a low powered device, one would think that passing the specific cflags would amount to some efficiency gains.

My distro default C(XX)FLAGS:
```
CFLAGS="-march=armv7-a -mfloat-abi=hard -mfpu=vfpv3-d16 -O2 -pipe -fstack-protector-strong -fno-plt"
CXXFLAGS="-march=armv7-a -mfloat-abi=hard -mfpu=vfpv3-d16 -O2 -pipe -fstack-protector-strong -fno-plt"
```

How I built:
```
  CFLAGS+=" -march=armv8-a+crc -mcpu=cortex-a53 -mfpu=neon-fp-armv8"
  CXXFLAGS="${CFLAGS}"

  cmake -DCMAKE_INSTALL_PREFIX=/usr \
    -DCMAKE_INSTALL_LIBDIR=/usr/lib \
    -DCMAKE_PREFIX_PATH=/opt/vc \
    -DCORE_PLATFORM_NAME=rbpi \
    -DWITH_CPU="cortex-a53" \
    -DENABLE_INTERNAL_FMT=ON \
    -DENABLE_EVENTCLIENTS=ON \
    -DENABLE_OPENGL=OFF \
    -DENABLE_X11=OFF \
    -DENABLE_VAAPI=OFF \
    -DENABLE_VDPAU=OFF \
    -DENABLE_INTERNAL_CROSSGUID=ON \
    -DLIRC_DEVICE=/run/lirc/lircd \
    -Dlibdvdcss_URL="/libdvdcss-1.4.1-Leia-Alpha-1.tar.gz" \
    -Dlibdvdnav_URL="/libdvdnav-6.0.0-Leia-Alpha-1.tar.gz" \
    -Dlibdvdread_URL="/libdvdread-6.0.0-Leia-Alpha-1.tar.gz" \
    -DFFMPEG_URL="/ffmpeg-3.4.1-Leia-Alpha-1.tar.gz" \
    ../"xbmc-18.0a1-Leia"
  make
  make preinstall

   _components=('kodi' 'kodi-bin')
   for _cmp in ${_components[@]}; do
     DESTDIR="$pkgdir" /usr/bin/cmake \
     -DCMAKE_INSTALL_COMPONENT="$_cmp" \
     -P cmake_install.cmake
   done
```

Environment: RPi3 on Arch ARMv7 (armv7h) running kodi-18.0a1.

## Types of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [X] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
Note, I will scan though these docs to see if the `-DWITH_CPU=xxx` is mentioned.
- [ ] I have updated the documentation accordingly
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

graysky2 committed:
add support for RPi3

wsnipex commented:
see https://github.com/xbmc/xbmc/pull/12937

graysky2 commented:
@wsnipex - Interesting... here I thought I had a novel idea :p

I do not know which would be "better" choice for `mfpu` ... `neon-fp-armv8` or `crypto-neon-fp-armv8` as @candrews proposed.  It's tough to find even a synthetic benchmark comparing the two.. I don't put too much into those since the real world work loads (ie kodi stuff, ffmpeg decoding, or doing CPU intensive plugin stuff) could be different.

In any case, I think one would want to modify both `cmake/scripts/linux/ArchSetup.cmake` and `tools/depends/configure.ac` to add support, not just the latter one in #12937.  For consistency, I do like the idea of adding RPi3 since RPi1 and RPi2 are in there.

asavah commented:
@graysky2 
crypto-neon-fp-armv8 is useless on the pi3,
for ARMv8-a crypto extensions (aes, sha*, pmull instructions) are _optional_, and pi3 a53 cores dont't have them,
crypto extesions became mandatory in ARMv8.1-a.
I'm not sure if it could cause any harm or instability, but forcing a fpu flag that is not supported at all by the SoC seems pointless to me.
neon-fp-armv8 lgtm

wsnipex commented:
since those flags are really RPI specific, we cannot force them on all platforms and devices.
Please enclose those in if(CORE_PLATFORM_NAME_LC STREQUAL rbpi)
and add reasonable(conservative) defaults for other socs.

graysky2 commented:
@asavah - Thanks for the info there.  Let's go without the crypto additions.
@wsnipex - Let me review your request and rebase/update.

graysky2 commented:
@wsnipex - I might be misunderstanding.... based on the comments by @asavah, the crypto extensions do not add anything.  I believe that the PR as-is doesn't force anything for other socs.


 the PR as-is doesn't include any

graysky2 commented:
@wsnipex - I modified and rebased.

* `cmake/scripts/linux/ArchSetup.cmake` now contains a minimal set of options thinking that other socs with that CPU would benefit... do we want to add back `-mfloat-abi=soft` or omit leaving the defaults in place?
* `tools/depends/configure.ac` contains the RPi1/2/3 specific options.

Testing this now.  Please comment.

wsnipex reviewed (approved):
thanks, looks good

graysky2 commented:
@wsnipex @asavah - Since the pi2 and pi3 specific stuff are wrapped up in `if(CORE_PLATFORM_NAME_LC STREQUAL rbpi)`, is there any reason why we shouldn't hardcode in `-march=armv7ve` and `-march=armv8-a+crc` respectfully?

Rechi commented on cmake/scripts/linux/ArchSetup.cmake:
```diff
@@ -36,6 +36,19 @@ else()
 endif()
 
+# temp until further cleanup is done
+# add Raspberry Pi 2 and 3 specific flags
```
Raspberry Pi 1 is missing
If yon don't have special flags at least don't fail with `Unknown CPU: arm1176jzf-s`

graysky2 replied:
@Rechi - I left RPi1 alone where the first check would catch it.  When I inspect the `gcc -c -Q -mcpu=native --help=target` on Arch ARMv6, I see we could could adjust these to `-mtune=armv6zk ` and you're right about the fail.  Will modify, rebase, and push shortly.

Rechi replied:
`message(SEND_ERROR "Unknown CPU: ${CPU}")` will cause to not generate the project if you want to build for Raspberry Pi 1.

graysky2 replied:
I didn't see your edit while I edited my edit :)  I was mistaken about the alternative config for RPi1 and edited my comment accordingly.  The rebased commit just removes the fail per your suggestion.  Please review and comment.

Rechi replied:
looks good

wsnipex commented:
there is no reason to set -march if -mcpu is already set.

jenkins build this please

popcornmix commented:
It doesn't break my build, so no objections here.

graysky2 commented:
@wsnipex -
> there is no reason to set -march if -mcpu is already set.

The [gcc 7.2 docs](https://gcc.gnu.org/onlinedocs/gcc-7.2.0/gcc/ARM-Options.html) state:

> -mcpu=name - This specifies the name of the target ARM processor. GCC uses this name to derive the name of the target ARM architecture (as if specified by -march) and the ARM processor type for which to tune for performance (as if specified by -mtune).

As I read that, the value of `-march=` and `-mtune=` would be derived from the value of `-mcpu=` we are hard coding here.  What I don't understand is how GCC could know which of the possible values we are intending particularly if we're building in a chroot on a lesser piece of hardware.  I would think it would default to the Cortex-A53 minimum `march=` option, no?  Perhaps I am misunderstanding it.

graysky2 commented:
> @graysky2 this needs a rebase

Is this what you guys needed?  I am not proficiency with git to rebase fixup my original PR with the resolved commit.

Rechi commented:
Yes we want a rebase, we don't allow PRs to contain merge commits, as they complicate the history
If you can't do it yourself, I can do it for you.

graysky2 commented:
@Rechi - If you wouldn't mind, please share the command line with me... I have rebased (fixup) consecutive commits before, but in this case, it looks like my original PR is 24 commits before the one I just did via the WebGUI.

Rechi commented:
I assume you have xbmc/xbmc named upstream and graysky2/xbmc origin and your PR branch checked out.

```
git reset --hard b6e8b5f # to revert to your initial state of the PR
git fetch upstream
git rebase upstream/master
# now correct the conflicts
git add -- cmake/scripts/linux/ArchSetup.cmake
git rebase --continue
git push -f origin master
```

graysky2 commented:
@Rechi - Thanks for the tutorial... my google-fu is lacking.  For the benefit of some other git-newbie finding this.  I had to issue the following before doing the steps Rechi outlined:
` git remote add upstream git@github.com:xbmc/xbmc.git` which inserted the following into my `.git/config`
```
[remote "upstream"]
  url = git@github.com:xbmc/xbmc.git
  fetch = +refs/heads/*:refs/remotes/upstream/*
```
